locality_sensitive_hashing/lsh.py

```diff
@@ -164,14 +166,81 @@ def k_shingles() -> KShingles:
         for i in range(len(doc) - k + 1):
             shingles.add(doc[i : i + k])
 
-        docs_k_shingles.append(shingles)
+        docs_k_shingles.append(np.array(list(shingles)))
     return docs_k_shingles
 
 
+def miller_rabin_test(n: int, k: int = 100) -> bool:
+    """
+    Miller-Rabin test
+    Derived from pseudocode in [Miller-Rabin test](https://en.wikipedia.org/wiki/Miller-Rabin_primality_test).
+
+    Arguments:
+        n: int
+            An odd integer > 2 to be tested for primality
+        k: int
+            The number of rounds of testing to perform
+        seed: int
+            The seed for the random number generator
+
+    Returns:
+        bool
+            True if n is _probably_ prime, False if n is _definitely_ composite
+    """
+    s, d = 0, n - 1
+    while d % 2 == 0:
+        s += 1
+        d //= 2
+
+    for _ in range(k):
+        a = random.randint(2, n - 1)
+        x = pow(a, d, n)
+
+        if x == 1 or x == n - 1:
+            continue
+
+        for _ in range(s - 1):
+            x = pow(x, 2, n)
+            if x == 1:
+                return False
+
+            if x == n - 1:
+                break
+        else:
+            return False
+    return True
+
+
+def next_prime(n: int, seed=42) -> int:
+    """
+    Finds the next probable prime number after `n`, based on the Miller-Rabin test.
+
+    Arguments:
+        n: int
+            An odd integer > 2
+    """
+    random.seed(seed)
+
+    assert n % 2 == 1, "n must be odd"
+    assert n > 2, "n must be greater than 2"
+
+    while not (miller_rabin_test(n)):
+        n += 2
+    return n
+
+
+total_shingles: int = 0
+
+
 # METHOD FOR TASK 2
 # Creates a signatures set of the documents from the k-shingles list
-def signature_set(k_shingles: KShingles) -> SignatureSet:
+def signature_set(k_shingles: KShingles) -> DenseSignatureSet:
     """
+    For a given list of k-shingles per document, creates a dense signature set of the documents.
+    Since a boolean matrix representation is [typically sparse](http://infolab.stanford.edu/~ullman/mining/2009/similarity1.pdf),
+    we will use a dense representation instead.
+
```

Comment:
Redundant double newline? (extreme nitpick)

Reply:
Haha, removed in 755a3f2